ReadMe
                       SQL-Routine-SQLBuilder
----------------------------------------------------------------------

by Darren Duncan <perl@DarrenDuncan.net>

CONTENTS

* Preface
* Keeping Up To Date
* Dependencies
* Support
* License And Copyright
* Acknowledgements

PREFACE

This distribution features the Perl 5 module named
"SQL::Routine::SQLBuilder"; see the file lib/SQL/Routine/SQLBuilder.pm for
the main documentation, which is in POD format.  The distribution also
includes a ChangeLog file to say what changes already happened, and a TODO
file to say what changes have yet to happen.  The file INSTALL should help
with installation issues.  Read below for how to get support or keep up to
date, know this distribution's dependencies, or read the credits or
copyright and license.

KEEPING UP TO DATE

My module set is constantly under development.  The canonical copies are
all linked to through my website, "http://www.DarrenDuncan.net/", on the
page called "Perl Libraries I Made" (name subject to change).

For those of you that are interested in the bleeding edge, I now have a
public Subversion repository at "http://svn.utsl.gen.nz/trunk/Rosetta/";
this repository contains all of my newer modules (whose problem domain is
databases), but not my older modules (which are not about databases).

You can download slightly older but more stable versions of all of my
modules in standard tarred and gzipped distributions, either from the
public CPAN archives, "http://www.cpan.org/authors/id/D/DU/DUNCAND/", or
from my own web server, "http://www.DarrenDuncan.net/d/perl/".

CPAN also extracts the POD from modules and maintains that documentation in
an organized and searchable fashion.  Their main search site is
"http://search.cpan.org/", and my own DUNCAND author page is
"http://search.cpan.org/~duncand/"; distributions are also there.

You can download my oldest distributions from BACKPAN, at
"http://history.perl.org/backpan/authors/id/D/DU/DUNCAND/", or from my web
server, at "http://www.DarrenDuncan.net/d/perl/archives/".

DEPENDENCIES

This distribution requires any version of Perl 5.x.y that is at least
5.8.1.

It also requires the Perl modules 'version' and 'only', which would
conceptually be built-in to Perl, but aren't, so they are on CPAN instead.

It also requires the Perl module List::MoreUtils '0.12-', which would
conceptually be built-in to Perl, but isn't, so it is on CPAN instead.

It also requires these modules that are on CPAN: Locale::KeyedText '1.6.0-'
(for error messages), SQL::Routine '0.70.0-'.

This distribution also requires Test::More for its test suite, and that
module is bundled with Perl.

SUPPORT

Currently I don't have any support arranged with other people, lists,
newsgroups, or otherwise.  Feel free to ask me if you can't figure things
out on your own, or another person whom you know has used this.  I may
start a mailing list for support issues later, so that users of my modules
can help each other with them.

LICENCE AND COPYRIGHT

This file is part of the SQL::Routine::SQLBuilder reference implementation
of a SQL:2003 string builder that uses the SQL::Routine database
portability library.

SQL::Routine::SQLBuilder is Copyright (c) 2002-2005, Darren R. Duncan.  All
rights reserved.  Address comments, suggestions, and bug reports to
perl@DarrenDuncan.net, or visit http://www.DarrenDuncan.net/ for more
information.

SQL::Routine::SQLBuilder is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License (GPL) as
published by the Free Software Foundation (http://www.fsf.org/); either
version 2 of the License, or (at your option) any later version.  You
should have received a copy of the GPL as part of the
SQL::Routine::SQLBuilder distribution, in the file named "GPL"; if not,
write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
Boston, MA  02110-1301, USA.

Linking SQL::Routine::SQLBuilder statically or dynamically with other
modules is making a combined work based on SQL::Routine::SQLBuilder.  Thus,
the terms and conditions of the GPL cover the whole combination.  As a
special exception, the copyright holders of SQL::Routine::SQLBuilder give
you permission to link SQL::Routine::SQLBuilder with independent modules,
regardless of the license terms of these independent modules, and to copy
and distribute the resulting combined work under terms of your choice,
provided that every copy of the combined work is accompanied by a complete
copy of the source code of SQL::Routine::SQLBuilder (the version of
SQL::Routine::SQLBuilder used to produce the combined work), being
distributed under the terms of the GPL plus this exception.  An independent
module is a module which is not derived from or based on
SQL::Routine::SQLBuilder, and which is fully useable when not linked to
SQL::Routine::SQLBuilder in any form.

Any versions of SQL::Routine::SQLBuilder that you modify and distribute
must carry prominent notices stating that you changed the files and the
date of any changes, in addition to preserving this original copyright
notice and other credits.  SQL::Routine::SQLBuilder is distributed in the
hope that it will be useful, but WITHOUT ANY WARRANTY; without even the
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

While it is by no means required, the copyright holders of
SQL::Routine::SQLBuilder would appreciate being informed any time you
create a modified version of SQL::Routine::SQLBuilder that you are willing
to distribute, because that is a practical way of suggesting improvements
to the standard version.

ACKNOWLEDGEMENTS

None besides myself so far.

FIN

Share and Enjoy!


Changes
                       SQL-Routine-SQLBuilder
----------------------------------------------------------------------

2005-09-28   Darren Duncan <perl@DarrenDuncan.net>

    * SQL-Routine-SQLBuilder Release 0.21.2 (SVN r878).

    * New code file versions are: SQLBuilder.pm 0.21.2 and en.pm 0.2.2.

    * From now on, any ordinary 'use [|only ]Foo' statements that appear
    within *.pm files will be located in different places than before; the
    'use' statements for modules that do not export anything, particularly
    object oriented modules, will appear near the top of the file, above
    all package declarations, but just below the use-pragma statements; the
    'use' statements for modules whose exported functions we are using will
    be placed just below the declarations of each package in which the
    relevant exported functions are used.

    * Removed any line-trailing whitespace from all distribution files; the
    one exception was those inside the SRT_SB_10_Generic.t here-docs.

    * Reformatted all code by swapping various string quoting delimiters.

    * Reformatted SQLBuilder.pm to un-cuddle all 'else' and 'elsif' blocks.

    * Updated SQLBuilder.pm and t_SRT_SB_Util.pm to replace all of their
    'unless' conditionals with equivalent 'if' conditionals.

    * Updated SQLBuilder.pm to rename all of its 'foreach' loops to 'for'.

    * Updated SQLBuilder.pm to reformat the 5 '|' delimited regular
    expressions as '/' delimited ones.

    * Updated SQLBuilder.pm to reformat all 5 regular expressions so that
    they use the /x flag.

    * Reformatted all code to change any "if( ... ) {" to "if (...) {".

    * Updated SQLBuilder.pm to reformat all 4 'while' like the 'if' were.

    * Added named constant $EMPTY_STR to SQLBuilder.pm.

    * Updated SQLBuilder.pm to add labels to the 1 un-labeled
    [next|last|redo] statement, and its innermost bounding loop control.

    * Reformatted all code so that any uses of the named unary operators
    [defined|ref] no longer have parenthesis around their argument.

    * Updated SQLBuilder.pm to reformat 1 use of 'exists', 2 uses of 'uc',
    and 1 use of 'lc' like with 'ref'.

    * Removed 1 superfluous use of 'scalar' in SQLBuilder.pm.

    * Updated SQLBuilder.pm to replace 2 "<condition> and return ...;" each
    with "return ... if <condition>;".

    * Updated SQLBuilder.pm to replace 2 "<condition> or $self->_throw ..."
    each with "$self->_throw ... if <reverse-condition>".

    * Updated t_SRT_SB_Util.pm to replace 1 "if ... return ..." with its
    postfix-if equivalent.

    * Updated SQLBuilder.pm to rearrange 4 conditional '_throw...' statements
    so they have postfix-if conditionals.

    * Reformatted all code, in SQLBuilder.pm, so that any uses of the
    built-in operators or functions named [bless|join|push|split|unpack] no
    longer have parenthesis around their argument lists; either the parens
    were just removed, or they were moved to surround both the
    operator/function name and its arguments.

    * Added new external dependency on the honorary built-in function
    List::MoreUtils::all() to SQLBuilder.pm.  Then updated the
    build_schema_or_app_table_create() method to use this function rather
    than a last-terminated for-loop when determining whether all columns in
    an index are mandatory.  Specifically, List::MoreUtils versions 0.12
    and up are required, since versions 0.11 and below didn't export some
    functions.

    * Split up some long lines in en.pm.

    * Updated any DEPENDENCIES documentation to reformat any indented
    sections as paragraphs, and show version ranges like 'only' takes them.

    * Reformatted all code so that every occurance of the string
    concatenation operator (.) has a space between the operator and each of
    its 2 arguments, rather than their all being in contact.  Also
    reformatted any lines that are split on this operator so the operator
    appears at the start of the second line, rather than the end of the
    first line.

    * Reformatted a few multi-line statements so that any [=|and|or] that
    are split on appear at the start of a line rather than the end of one.

    * Reformatted all of the code comments at the top of SQLBuilder.pm,
    where its object property names, and constant values, are declared,
    so that all comment lines don't exceed the 75 character line length.

    * In SQLBuilder.pm, reformatted 18 cascading ternary statements into
    aligned columns.

    * Other miscellaneous code line alignments and splitting.

2005-09-12   Darren Duncan <perl@DarrenDuncan.net>

    * SQL-Routine-SQLBuilder Release 0.21.1 (SVN r808).

    * New code file versions are: SQLBuilder.pm 0.21.1 and en.pm 0.2.1.

    * Updated all POD-containing files to re-wrap any non-indented POD
    paragraphs to a 75 character width, which is 5 less than the 80
    character width they were wrapped to before.  This change should make
    it easier to copy and paste a diff or patch of this documentation into
    an email message, where it is quoted at least once, without any line
    wrapping occurring.  Likewise, these standard documentation files were
    re-wrapped to 75 characters: ReadMe, INSTALL, LGPL|GPL, Changes, TODO.

    * Revised this Changes file to replace all detail entries for releases
    0.18 thru 0.21.0 with a significant release list for the same period;
    also added references to when all 'Rosetta developer release' (#s 1-3)
    were.

2005-09-08   Darren Duncan <perl@DarrenDuncan.net>

    * SQL-Routine-SQLBuilder Release 0.21.0 (SVN r787), containing
    SQL::Routine::SQLBuilder 0.21.0, was released on CPAN.  This is the
    first release of this distribution and its modules for which they had
    3-part version numbers, rather than floating point version numbers.

    * These were the current versions cited by the public announcement for
    Rosetta/SQL-Routine developer release #3.


t/lib/t_SRT_SB_Util.pm
#!perl
use 5.008001; use utf8; use strict; use warnings;

# This module is used when testing SQL::Routine::SQLBuilder.
# It contains some utility methods used by the various SRT_SB_*.t scripts.

package # hide this class name from PAUSE indexer
t_SRT_SB_Util;

######################################################################

sub message {
    my (undef, $detail) = @_;
    print "# $detail\n";
}

######################################################################

sub error_to_string {
    my (undef, $message) = @_;
    if (ref $message and UNIVERSAL::isa( $message, 'Locale::KeyedText::Message' )) {
        my $translator = Locale::KeyedText->new_translator(
            ['SQL::Routine::SQLBuilder::L::', 'SQL::Routine::L::'], ['en'] );
        my $user_text = $translator->translate_message( $message );
        return q{internal error: can't find user text for a message: }
            . $message->as_string() . ' ' . $translator->as_string()
            if !$user_text;
        return $user_text;
    }
    return $message; # if this isn't the right kind of object
}

######################################################################

1;


t/lib/t_SRT_SB_Model.pm
#!perl
use 5.008001; use utf8; use strict; use warnings;

# This module is used when testing SQL::Routine::SQLBuilder.

package # hide this class name from PAUSE indexer
t_SRT_SB_Model;

######################################################################

sub populate_model {
    my (undef, $model) = @_;

    $model->build_child_node_trees( [
        [ 'scalar_data_type', { 'si_name' => 'entity_id'  , 'base_type' => 'NUM_INT' , 'num_precision' => 9, }, ],
        [ 'scalar_data_type', { 'si_name' => 'alt_id'     , 'base_type' => 'STR_CHAR', 'max_chars' => 20, 'char_enc' => 'UTF8', }, ],
        [ 'scalar_data_type', { 'si_name' => 'person_name', 'base_type' => 'STR_CHAR', 'max_chars' => 100, 'char_enc' => 'UTF8', }, ],
        [ 'scalar_data_type', { 'si_name' => 'person_sex' , 'base_type' => 'STR_CHAR', 'max_chars' => 1, 'char_enc' => 'UTF8', }, [
            [ 'scalar_data_type_opt', 'M', ],
            [ 'scalar_data_type_opt', 'F', ],
        ], ],
        [ 'row_data_type', 'person', [
            [ 'row_data_type_field', { 'si_name' => 'person_id'   , 'scalar_data_type' => 'entity_id'  , }, ],
            [ 'row_data_type_field', { 'si_name' => 'alternate_id', 'scalar_data_type' => 'alt_id'     , }, ],
            [ 'row_data_type_field', { 'si_name' => 'name'        , 'scalar_data_type' => 'person_name', }, ],
            [ 'row_data_type_field', { 'si_name' => 'sex'         , 'scalar_data_type' => 'person_sex' , }, ],
            [ 'row_data_type_field', { 'si_name' => 'father_id'   , 'scalar_data_type' => 'entity_id'  , }, ],
            [ 'row_data_type_field', { 'si_name' => 'mother_id'   , 'scalar_data_type' => 'entity_id'  , }, ],
        ], ],
        [ 'row_data_type', 'person_with_parents', [
            [ 'row_data_type_field', { 'si_name' => 'self_id'    , 'scalar_data_type' => 'entity_id'  , }, ],
            [ 'row_data_type_field', { 'si_name' => 'self_name'  , 'scalar_data_type' => 'person_name', }, ],
            [ 'row_data_type_field', { 'si_name' => 'father_id'  , 'scalar_data_type' => 'entity_id'  , }, ],
            [ 'row_data_type_field', { 'si_name' => 'father_name', 'scalar_data_type' => 'person_name', }, ],
            [ 'row_data_type_field', { 'si_name' => 'mother_id'  , 'scalar_data_type' => 'entity_id'  , }, ],
            [ 'row_data_type_field', { 'si_name' => 'mother_name', 'scalar_data_type' => 'person_name', }, ],
        ], ],
    ] );

    my $catalog_bp = $model->build_child_node_tree( 'catalog', 'Gene Database', [
        [ 'owner', 'Lord of the Root', ],
        [ 'schema', { 'si_name' => 'Gene Schema', 'owner' => 'Lord of the Root', }, ],
    ] );
    my $schema = $catalog_bp->find_child_node_by_surrogate_id( 'Gene Schema' );

    my $tb_person = $schema->build_child_node_tree( 'table', { 'si_name' => 'person', 'row_data_type' => 'person', }, [
        [ 'table_field', { 'si_row_field' => 'person_id', 'mandatory' => 1, 'default_val' => 1, 'auto_inc' => 1, }, ],
        [ 'table_field', { 'si_row_field' => 'name'     , 'mandatory' => 1, }, ],
        [ 'table_index', { 'si_name' => 'primary' , 'index_type' => 'UNIQUE', }, [
            [ 'table_index_field', 'person_id', ],
        ], ],
        [ 'table_index', { 'si_name' => 'ak_alternate_id', 'index_type' => 'UNIQUE', }, [
            [ 'table_index_field', 'alternate_id', ],
        ], ],
        [ 'table_index', { 'si_name' => 'fk_father', 'index_type' => 'FOREIGN', 'f_table' => 'person', }, [
            [ 'table_index_field', { 'si_field' => 'father_id', 'f_field' => 'person_id' } ],
        ], ],
        [ 'table_index', { 'si_name' => 'fk_mother', 'index_type' => 'FOREIGN', 'f_table' => 'person', }, [
            [ 'table_index_field', { 'si_field' => 'mother_id', 'f_field' => 'person_id' } ],
        ], ],
    ] );

    my $vw_pwp = $schema->build_child_node_tree( 'view', { 'si_name' => 'person_with_parents',
            'view_type' => 'JOINED', 'row_data_type' => 'person_with_parents', }, [
        ( map { [ 'view_src', { 'si_name' => $_, 'match' => 'person', }, [
            map { [ 'view_src_field', $_, ], } ( 'person_id', 'name', 'father_id', 'mother_id', ),
        ], ], } ('self') ),
        ( map { [ 'view_src', { 'si_name' => $_, 'match' => 'person', }, [
            map { [ 'view_src_field', $_, ], } ( 'person_id', 'name', ),
        ], ], } ( 'father', 'mother', ) ),
        [ 'view_field', { 'si_row_field' => 'self_id'    , 'src_field' => ['person_id','self'  ], }, ],
        [ 'view_field', { 'si_row_field' => 'self_name'  , 'src_field' => ['name'     ,'self'  ], }, ],
        [ 'view_field', { 'si_row_field' => 'father_id'  , 'src_field' => ['person_id','father'], }, ],
        [ 'view_field', { 'si_row_field' => 'father_name', 'src_field' => ['name'     ,'father'], }, ],
        [ 'view_field', { 'si_row_field' => 'mother_id'  , 'src_field' => ['person_id','mother'], }, ],
        [ 'view_field', { 'si_row_field' => 'mother_name', 'src_field' => ['name'     ,'mother'], }, ],
        [ 'view_join', { 'lhs_src' => 'self', 'rhs_src' => 'father', 'join_op' => 'LEFT', }, [
            [ 'view_join_field', { 'lhs_src_field' => 'father_id', 'rhs_src_field' => 'person_id' } ],
        ], ],
        [ 'view_join', { 'lhs_src' => 'self', 'rhs_src' => 'mother', 'join_op' => 'LEFT', }, [
            [ 'view_join_field', { 'lhs_src_field' => 'mother_id', 'rhs_src_field' => 'person_id' } ],
        ], ],
    ] );

    my $application_bp = $model->build_child_node_tree( 'application', 'Gene App', [
        [ 'catalog_link', { 'si_name' => 'editor_link', 'target' => $catalog_bp, }, ],
    ] );

    $model->build_child_node( 'application_instance', { 'si_name' => 'test app', 'blueprint' => $application_bp, } );
}

######################################################################

1;


lib/SQL/Routine/SQLBuilder/L/en.pm
#!perl
use 5.008001; use utf8; use strict; use warnings;

package SQL::Routine::SQLBuilder::L::en;
use version; our $VERSION = qv('0.2.2');

######################################################################

my $CSB = 'SQL::Routine::SQLBuilder';

my %text_strings = (
    'SRT_SB_METH_ARG_UNDEF' =>
        $CSB . q[.{METH}(): ]
        . q[undefined (or missing) {ARGNM} argument],
    'SRT_SB_METH_ARG_NO_ARY' =>
        $CSB . q[.{METH}(): ]
        . q[invalid {ARGNM} argument; ]
        . q[it is not a Array ref, but rather is "{ARGVL}"],
    'SRT_SB_METH_ARG_NO_HASH' =>
        $CSB . q[.{METH}(): ]
        . q[invalid {ARGNM} argument; ]
        . q[it is not a Hash ref, but rather is "{ARGVL}"],
    'SRT_SB_METH_ARG_NO_NODE' =>
        $CSB . q[.{METH}(): ]
        . q[invalid {ARGNM} argument; ]
        . q[it is not a Node object, but rather is "{ARGVL}"],

    'SRT_SB_METH_ARG_WRONG_NODE_TYPE' =>
        $CSB . q[.{METH}(): ]
        . q[invalid {ARGNM} argument; ]
        . q[it is not a "{EXPNTYPE}" Node, but rather is a "{ARGNTYPE}" Node],

    'SRT_SB_IDENT_STYLE_ARG_INVAL' =>
        $CSB . q[.identifier_style(): ]
        . q[invalid NEW_VALUE argument; ]
        . q[it is not one of "{EXPVLS}", but rather is "{ARGVL}"],
);

######################################################################

sub get_text_by_key {
    my (undef, $msg_key) = @_;
    return $text_strings{$msg_key};
}

######################################################################

1;
__END__

=encoding utf8

=head1 NAME

SQL::Routine::SQLBuilder::L::en - Localization of SQL::Routine::SQLBuilder for English

=head1 VERSION

This document describes SQL::Routine::SQLBuilder::L::en version 0.2.2.

=head1 SYNOPSIS

    use Locale::KeyedText;
    use SQL::Routine::SQLBuilder;

    # do work ...

    my $translator = Locale::KeyedText->new_translator(
        ['SQL::Routine::SQLBuilder::L::', 'SQL::Routine::L::'], ['en'] );

    # do work ...

    eval {
        # do work with SQL::Routine::SQLBuilder, which may throw an exception ...
    };
    if (my $error_message_object = $@) {
        # examine object here if you want and programmatically recover...

        # or otherwise do the next few lines...
        my $error_user_text = $translator->translate_message( $error_message_object );
        # display $error_user_text to user by some appropriate means
    }

    # continue working, which may involve using SQL::Routine::SQLBuilder some more ...

=head1 DESCRIPTION

The SQL::Routine::SQLBuilder::L::en Perl 5 module contains localization
data for SQL::Routine::SQLBuilder.  It is designed to be interpreted by
Locale::KeyedText.  Besides localizing generic error messages that
SQL::Routine::SQLBuilder produces itself, this file also provides a
ready-made set of generic database error strings that can be thrown by any
SQL::Routine::SQLBuilder Engine.

This class is optional and you can still use SQL::Routine::SQLBuilder
effectively without it, especially if you plan to either show users
different error messages than this class defines, or not show them anything
because you are "handling it".

=head1 FUNCTIONS

=head2 get_text_by_key( MSG_KEY )

    my $user_text_template = SQL::Routine::SQLBuilder::L::en->get_text_by_key( 'foo' );

This function takes a Message Key string in MSG_KEY and returns the
associated user text template string, if there is one, or undef if not.

=head1 DEPENDENCIES

This module requires any version of Perl 5.x.y that is at least 5.8.1.

It also requires the Perl module L<version>, which would conceptually be
built-in to Perl, but isn't, so it is on CPAN instead.

This module has no enforced dependencies on L<Locale::KeyedText>, which is
on CPAN, or on L<SQL::Routine::SQLBuilder>, which is in the current
distribution, but it is designed to be used in conjunction with them.

=head1 INCOMPATIBILITIES

None reported.

=head1 SEE ALSO

L<perl(1)>, L<Locale::KeyedText>, L<SQL::Routine::SQLBuilder>.

=head1 BUGS AND LIMITATIONS

The structure of this module is trivially simple and has no known bugs.

However, the locale data that this module contains may be subject to large
changes in the future; you can determine the likeliness of this by
examining the development status and/or BUGS AND LIMITATIONS documentation
of the other module that this one is localizing; there tends to be a high
correlation in the rate of change between that module and this one.

=head1 AUTHOR

Darren R. Duncan (C<perl@DarrenDuncan.net>)

=head1 LICENCE AND COPYRIGHT

This file is part of the SQL::Routine::SQLBuilder reference implementation
of a SQL:2003 string builder that uses the SQL::Routine database
portability library.

SQL::Routine::SQLBuilder is Copyright (c) 2002-2005, Darren R. Duncan.  All
rights reserved.  Address comments, suggestions, and bug reports to
C<perl@DarrenDuncan.net>, or visit L<http://www.DarrenDuncan.net/> for more
information.

SQL::Routine::SQLBuilder is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License (GPL) as
published by the Free Software Foundation (L<http://www.fsf.org/>); either
version 2 of the License, or (at your option) any later version.  You
should have received a copy of the GPL as part of the
SQL::Routine::SQLBuilder distribution, in the file named "GPL"; if not,
write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
Boston, MA  02110-1301, USA.

Linking SQL::Routine::SQLBuilder statically or dynamically with other
modules is making a combined work based on SQL::Routine::SQLBuilder.  Thus,
the terms and conditions of the GPL cover the whole combination.  As a
special exception, the copyright holders of SQL::Routine::SQLBuilder give
you permission to link SQL::Routine::SQLBuilder with independent modules,
regardless of the license terms of these independent modules, and to copy
and distribute the resulting combined work under terms of your choice,
provided that every copy of the combined work is accompanied by a complete
copy of the source code of SQL::Routine::SQLBuilder (the version of
SQL::Routine::SQLBuilder used to produce the combined work), being
distributed under the terms of the GPL plus this exception.  An independent
module is a module which is not derived from or based on
SQL::Routine::SQLBuilder, and which is fully useable when not linked to
SQL::Routine::SQLBuilder in any form.

Any versions of SQL::Routine::SQLBuilder that you modify and distribute
must carry prominent notices stating that you changed the files and the
date of any changes, in addition to preserving this original copyright
notice and other credits.  SQL::Routine::SQLBuilder is distributed in the
hope that it will be useful, but WITHOUT ANY WARRANTY; without even the
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

While it is by no means required, the copyright holders of
SQL::Routine::SQLBuilder would appreciate being informed any time you
create a modified version of SQL::Routine::SQLBuilder that you are willing
to distribute, because that is a practical way of suggesting improvements
to the standard version.

=cut
